### PR TITLE
Fix language files for Node+ESM

### LINKED ISSUE #1344: Language files do not work with ES modules in node
### Description

Importing language files:
- seems to work in UMD module system
- seems to work in CommonJS although we should verify if the setup is optimal
- does not work in ESM in node
- does not work in ESM in React

### Demo

https://codesandbox.io/p/devbox/test-hyperformula-with-nodejs-pgkwd8

### Error messages
```
Error [ERR_MODULE_NOT_FOUND]: Cannot find module '/workspaces/sandbox/node_modules/hyperformula/es/i18n/languages/frFR' imported from /workspaces/sandbox/index.js
Did you mean to import hyperformula/es/i18n/languages/frFR.js?
```

```
(node:328) Warning: To load an ES module, set "type": "module" in the package.json or use the .mjs extension.
(Use `node --trace-warnings ...` to show where the warning was created)
/workspaces/sandbox/node_modules/hyperformula/es/i18n/languages/frFR.js:417
export default dictionary;
^^^^^^

SyntaxError: Unexpected token 'export'
```

### Notes

- Originally reported by @guenlm in https://github.com/handsontable/hyperformula/discussions/1340
- Consulted with @budnix in https://handsoncode.slack.com/archives/D0300KBP0H2/p1701186053349819 (internal discussion)
- The fix might require a breaking change.

### Related issues
 
- https://github.com/handsontable/hyperformula/issues/1143
- https://github.com/handsontable/hyperformula/issues/1144

### PR DESCRIPTION
**Pull request by sequba**
### Context

- changed babel configuration to produce the ESM build with `.mjs` extensions and adjust the import and export statements in the output (custom babel plugin)
- added `exports` property to `package.json` that lists all valid ways of importing HyperFormula
- added the migration guide

Useful reading about similar solutions:
- https://www.sensedeep.com/blog/posts/2021/how-to-create-single-source-npm-module.html
- https://github.com/microsoft/TypeScript/issues/49462
- https://github.com/microsoft/TypeScript/issues/18442

### Related issues:
Fixes #1344 

### New import paths for the language files

<table>
<tr>
  <th>Module system</th>
  <th>valid import paths in HF 2.x</th>
  <th>valid import paths in HF 3.0.0</th>
</tr>
<tr>
  <td>ES modules</td>
  <td><ul>
    <li>hyperformula/es/i18n/languages/</li>
    <li>hyperformula/es/i18n/languages/frFR</li>
    <li>hyperformula/es/i18n/languages/frFR,js</li>
  </ul></td>
  <td><ul>
    <li><strong>hyperformula/i18n/languages/</strong></li>
    <li><strong>hyperformula/i18n/languages/frFR</strong></li>
    <li>hyperformula/es/i18n/languages/ (legacy)</li>
    <li>hyperformula/es/i18n/languages/frFR (legacy)</li>
    <li>hyperformula/es/i18n/languages/frFR,js (legacy)</li>
  </ul></td></tr>
<tr>
  <td>CommonJS</td>
  <td><ul>
    <li>hyperformula/commonjs/i18n/languages/</li>
    <li>hyperformula/commonjs/i18n/languages/frFR</li>
    <li>hyperformula/commonjs/i18n/languages/frFR.js</li>
  </ul></td>
  <td><ul>
    <li><strong>hyperformula/i18n/languages/</strong></li>
    <li><strong>hyperformula/i18n/languages/frFR</strong></li>
    <li>hyperformula/commonjs/i18n/languages/ (legacy)</li>
    <li>hyperformula/commonjs/i18n/languages/frFR (legacy)</li>
    <li>hyperformula/commonjs/i18n/languages/frFR.js (legacy)</li>
  </ul></td>
</tr>
<tr>
  <td>UMD</td>
  <td>hyperformula/dist/languages/</td>
  <td>hyperformula/dist/languages/</td>
</tr>
</table>

#### Angular

| Angular version | supported by Angular | legacy import paths | new import paths |
| - | - | - | - |
| 18                   | yes | :white_check_mark: | :white_check_mark:                                                                   |
| 17                   | yes | ⚠️works after configuring `moduleResolution`  | ⚠️works after configuring `moduleResolution`                               |
| 16 with Typescript 5 | until 2024-11-08 | ⚠️works after configuring `moduleResolution`  | ⚠️works after configuring `moduleResolution`                               |
| 16 with Typescript 4 | until 2024-11-08 | ⚠️works after upgrading to typescript 5 and configuring `moduleResolution` |⚠️works after upgrading to typescript 5 and configuring `moduleResolution` |
| 15                   | no | :x: | :x: |

#### Typescript

| Typescript version | legacy import paths | new import paths |
| - | - | - | 
| 5 | ⚠️works after configuring `moduleResolution` | ⚠️works after configuring `moduleResolution` |
| 4 | ⚠️works after configuring `moduleResolution` | ⚠️works after configuring `moduleResolution` |

#### Webpack

| Webpack version | legacy import paths | new import paths |
| - | - | - | 
| 5 | :white_check_mark: | :white_check_mark: |
| 5 + babel | :white_check_mark: | :white_check_mark: |
| 4 | ⚠️works after configuring `resolver-default` | :x:  |
| 4 + babel | ⚠️works after configuring `resolver-default` | :x:  |

Webpack 4 has 2 separate issues:
1. It doesn't know how to handle `mjs` files.
2. It doesn't support `exports` property to `package.json`.

For the former issue, there is a workaround. It's enough to configure a rule for `mjs` files. But it seems that there is no way to make it understand the `exports` property. For that reason, Webpack 4 can work with HyperFormula 3.0.0, but only with import paths that match the actual paths (ignoring the `exports` property that provides the artificial paths).

Further reading:
- https://github.com/webpack/webpack/issues/9509#issuecomment-1310419920
- https://stackoverflow.com/a/74957466

#### Parcel

| Parcel version | legacy import paths | new import paths |
| - | - | - | 
| <2.9.0 | :white_check_mark: | :x:                                                                    |
| >=2.9.0 | :white_check_mark: | ⚠️works after configuring `resolver-default`                                                     |

Further reading:
- https://parceljs.org/blog/v2-9-0/#new-resolver
- https://github.com/parcel-bundler/parcel/issues/4155

#### Test importing HF in various setups
- [x] ESM in node project
- [x] ESM in react project (with Typescript)
- [x] ESM in angular 18 project -> works with default configuration 
- [x] ESM in angular 17 project -> works after configuring `moduleResolution`
- [x] ESM in angular 16 project (typescript 5) -> works after configuring `moduleResolution`
- [x] ESM in angular 16 project (typescript 4) -> works after upgrading to typescript 5 and configuring `moduleResolution`
- [x] ESM in webpack 5 project
- [x] ESM in webpack 5 project + babel
- [x] ESM in webpack 4 project -> works only with legacy paths and requires special configuration
- [x] ESM in webpack 4 project + babel -> works only with legacy paths and requires special configuration
- [x] ESM in webpack 5 project with Typescript 5 -> works after configuring `moduleResolution`
- [x] ESM in webpack 5 project with Typescript 4 -> works after configuring `moduleResolution`
- [x] ESM in parcel 2.11 project -> works after configuring `resolver-default`
- [x] ESM in parcel 2.8 project -> only legacy paths
- [x] CJS in node project
- [x] CJS in parcel 2.11 project 
- [x] UMD in browser

Demo projects are available in the [hyperformula-demos repo](https://github.com/handsontable/hyperformula-demos/tree/import-demos).

#### Tested by @budnix:

> Vite ^5.1.4
> Parcel2 ^2.0.0
> Webpack5 ^5.8.0
> Webpack4 ^4.42.0

#### Tested by @magierg 

> Webpack 5.9 + TS
> React 18^ + TS
> Angular 16.0.2 + TS 4.4.3
> Angular 17.3.1 + TS 5.2
> Svelte 3.5.4 + vite 4
> Vue 3.3.4 + vite 4.4.9

#### Further verification

- [x] go through the angular setup and polish the guide
- [x] go through the typescript 4 and 5 setup and polish the guide
- [x] go through the webpack 4 and 5 setup and polish the guide
- [x] go through the parcel setup and polish the guide
- [x] go through all the notes and comments in the PR and make sure the guide is complete
- [x] validate by publint
- [x] test with https://arethetypeswrong.github.io/
- [x] CR
- [ ] (@magierg) test the migration guide: create new projects and set them up using the instructions from the guide
- [ ] (@magierg) final QA approval

### Types of changes

- [x] Breaking change (a fix or a feature because of which an existing functionality doesn't work as expected anymore)
- [x] New feature or improvement (a non-breaking change that adds functionality)
- [x] Bug fix (a non-breaking change that fixes an issue)
- [ ] Additional language file, or a change to an existing language file (translations)
- [ ] Change to the documentation

### Checklist:

- [x] I have reviewed the guidelines about [Contributing to HyperFormula](https://hyperformula.handsontable.com/guide/contributing.html) and I confirm that my code follows the code style of this project.
- [ ] I have signed the [Contributor License Agreement](https://goo.gl/forms/yuutGuN0RjsikVpM2).
- [x] My change is compliant with the [OpenDocument](https://docs.oasis-open.org/office/OpenDocument/v1.3/os/part4-formula/OpenDocument-v1.3-os-part4-formula.html) standard.
- [x] My change is compatible with Microsoft Excel.
- [x] My change is compatible with Google Sheets.
- [x] I described my changes in the [CHANGELOG.md](https://github.com/handsontable/hyperformula/blob/master/CHANGELOG.md) file.
- [ ] My changes require a documentation update.
- [x] My changes require a migration guide.

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -154,8 +159,191 @@
...
+    "./es/i18n/languages/daDK": {
+      "types": "./typings/i18n/languages/daDK.d.ts",
+      "import": "./es/i18n/languages/daDK.mjs",
+      "require": "./commonjs/i18n/languages/daDK.js"
+    },
+    "./commonjs/i18n/languages/daDK": {
+      "types": "./typings/i18n/languages/daDK.d.ts",
+      "import": "./es/i18n/languages/daDK.mjs",
+      "require": "./commonjs/i18n/languages/daDK.js"
+    },
```
Comment (budnix) [inline package.json:197]:
```suggestion
    "./i18n/languages/csCZ": {
      "types": "./typings/i18n/languages/csCZ.d.ts",
      "import": "./es/i18n/languages/csCZ.mjs",
      "require": "./commonjs/i18n/languages/csCZ.js"
    },
    "./i18n/languages/daDK": {
      "types": "./typings/i18n/languages/daDK.d.ts",
      "import": "./es/i18n/languages/daDK.mjs",
      "require": "./commonjs/i18n/languages/daDK.js"
    },
```
All these language entries should be merged into one so the user can use one path to import the dictionary.

**Review by budnix — CHANGES_REQUESTED** [review]
I tested the dist files in some more popular bundlers and here are the results:
 * Vite ^5.1.4 - OK
 * Parcel2 ^2.0.0 - Does not support "exports" so the language file has to be imported with `/es/` or `/commonjs/` prefixes;
 * Webpack5 ^5.8.0 - OK
 * Webpack4 ^4.42.0 - I getting an error. Seems the webpack has the problem with importing values from "chevrotain". 
```ts
// /node_modules/hyperformula/es/parser/FormulaParser.mjs
import { EmbeddedActionsParser, EMPTY_ALT, Lexer, tokenMatcher } from 'chevrotain';
// EmbeddedActionsParser, EMPTY_ALT and others are missing
```

```log
ERROR in ./node_modules/hyperformula/es/parser/FormulaParser.mjs 689:125-137
Can't import the named export 'tokenMatcher' from non EcmaScript module (only default export is available)
 @ ./node_modules/hyperformula/es/parser/index.mjs
 @ ./node_modules/hyperformula/es/ArraySize.mjs
 @ ./node_modules/hyperformula/es/index.mjs
 @ ./src/index.js

ERROR in ./node_modules/hyperformula/es/parser/FormulaParser.mjs 708:29-41
Can't import the named export 'tokenMatcher' from non EcmaScript module (only default export is available)
 @ ./node_modules/hyperformula/es/parser/index.mjs
 @ ./node_modules/hyperformula/es/ArraySize.mjs
 @ ./node_modules/hyperformula/es/index.mjs
 @ ./src/index.js
```

File: package.json (hunk 1 above)
Comment (sequba) [reply package.json:197]: Alright, I can add paths `./i18n/languages/xxXX`, but I also want to keep the old paths `./commonjs/...` and `./es/...` for backwards compatibility.

Comment (sequba) [reply package.json]: done

**Comment by magierg** [comment]
I have retested all the demos available on https://github.com/handsontable/hyperformula-demos/tree/import-demos 
image

plus the bundlers mentioned by @budnix getting the same results. 

On top of the above I have tested:

- Webpack 5.9 + TS - tsconfig.js setting change required:
``` 
"module": "node16",
"moduleResolution": "node16",
```
- React 18^ + TS - same tsconfig.js change required
- Angular 16.0.2 + TS 4.4.3 - @sequba to investigate tsconfig changes required
- Angular 17.3.1 + TS 5.2 - same issue with TS
- Svelte 3.5.4 + vite 4 - getting this error  - might be user error - @sequba to investigate
```
Error: Language already registered.
    at HyperFormula.registerLanguage (/Users/magierg/repos/hyperformula-demos/svelte-demo/node_modules/hyperformula/commonjs/HyperFormula.js:287:13)
    at Hyperformula.svelte:18:15
    at Object.$$render (/node_modules/svelte/internal/index.mjs:1892:22)
    at eval (/src/routes/+page.svelte:24:148)
    at Object.$$render (/node_modules/svelte/internal/index.mjs:1892:22)
    at Object.default (root.svelte:41:38)
    at eval (/node_modules/@sveltejs/kit/src/runtime/components/layout.svelte:8:41)
    at Object.$$render (/node_modules/svelte/internal/index.mjs:1892:22)
    at root.svelte:40:37
    at $$render (/node_modules/svelte/internal/index.mjs:1892:22)
```
- Vue 3.3.4 + vite 4.4.9 - OK

**Comment by sequba** [comment]
TODO: verify it with the code example in issue https://github.com/handsontable/hyperformula/issues/1143

**Comment by sequba** [comment]
> Svelte 3.5.4 + vite 4 - getting this error  - might be user error - @sequba to investigate

It works correctly. The issue you reported was caused by running the setup code twice. I created a demo for svelte: https://github.com/handsontable/hyperformula-demos/tree/import-demos/import-demo-esm-svelte

Angular case is still to be verified.

**Comment by sequba** [comment]
> Angular 16.0.2 + TS 4.4.3 - @sequba to investigate tsconfig changes required
> Angular 17.3.1 + TS 5.2 - same issue with TS

Angular issue can be resolved by setting `"moduleResolution": "bundler"` in `tsconfig.json`. Tested with Angular 17 and 18. [Demo](https://github.com/handsontable/hyperformula-demos/tree/import-demos/import-demo-esm-angular)

**Review by budnix — APPROVED** [review]
Looks good 👌

**Comment by magierg** [comment]
I have retested the previously mentioned configurations and performed the migration steps - found out that the ` @babel/core` is not required - this has been already addressed by @sequba

**Pull request merged**